File: prototypes/lif.m
clear;

VARIABLE_CURRENT = true;


T = 100; %total timesteps in simulation
tstep = 1; %timestep size
refractory_length = 10; %waiting period after a spike before integrating further
times = 0:tstep:T; %Possible times in the simulation


Rmem = 1; %The membrane resistance
Cmem = 10; %The membrande capacitance
tau_m = Rmem*Cmem; %The time constant
Vthresh = 2; %The Threshold voltage at which the neuron spikes
V_spike = .5; %How much added voltage once the neuron spikes

%Current inputs
Ibias = 2.2;
if (VARIABLE_CURRENT == true)
    periods = randsample(.1:.1:2,5)
    f1 = @(x) sin(periods(1) * x);
    f2 = @(x) sin(periods(2) * x);
    f3 = @(x) sin(periods(3) * x);
    f4 = @(x) sin(periods(4) * x);
    f5 = @(x) sin(periods(5) * x);

    I = @(x) (f1(x) + f2(x) + f3(x) + f4(x) + f5(x)) + Ibias;   
else
    I(1:length(times)) = Ibias; 
end



Vm = zeros(1,length(times));
t_rest = 0;
for i = 1:1:length(times)
    t = times(i);
    
    if t > t_rest
        Vm(i) = Vm(i-1) + (-Vm(i-1) + I(i)*Rmem) / tau_m * tstep; 
    end
    if Vm(i) >= Vthresh
        Vm(i) = Vm(i) + V_spike;
        t_rest = t + refractory_length;
    end
    
    
end

plot(times, Vm)

File: prototypes/lif.py
import sys
import numpy as np
import matplotlib.pyplot as plt

VARIABLE_CURRENT = False
T = 100
tstep = 1
refractory_length = 10
times = np.arange(0,T,tstep)

Rmem = 1
Cmem = 10
tau_m = Rmem*Cmem
Vthresh = 2
V_spike = .5

# Current Inputs
def myfun(x):
    Ibias = 2.2
    periods = np.arange(.1,2,.1)
    np.random.shuffle(periods)

    funperiods = []
    for i in range(0,5):
        funperiods.append(periods[i])
        # funperiods[i] = periods[i]

    output = 0
    for P in funperiods:
        output += np.sin(P*x)

    return output + Ibias



if __name__ == '__main__':
    if sys.argv[1] == '-v':
        VARIABLE_CURRENT = True

    Vm = np.zeros(len(times))
    t_rest = 0
    for i in range(0, len(times)):
        t = times[i]

        if t > t_rest:
            Vm[i] = Vm[i-1] + (-Vm[i-1] + myfun(t)*Rmem) / tau_m * tstep
        if Vm[i] >= Vthresh:
            Vm[i] = Vm[i] + V_spike
            t_rest = t + refractory_length

    plt.plot(times, Vm)
    plt.show()
